src/plugins/vis_types/vega/public/vega_view/vega_base_view.js

```diff
@@ -44,6 +45,47 @@ for (const funcName of Object.keys(vegaFunctions)) {
 
 const bypassToken = Symbol();
 
+function normalizeDate(date) {
+  let normalizedDate = '';
+  if (typeof date === 'number') {
+    normalizedDate = date;
+  } else if (date instanceof Date) {
+    normalizedDate = date;
+  } else {
+    normalizedDate = normalizeString(date);
+  }
+  return normalizedDate;
+}
+
+// recursive function to check a nested object for the 'toJSON' property
+function checkObjectForToJsonProperty(object) {
+  if (object === null || object === undefined) {
+    return false;
+  }
+  if (typeof object === 'object' && 'toJSON' in object) {
+    return true;
+  }
+  const hasProperty = Object.keys(object).some((key) => {
+    if (typeof object[key] === 'object') {
+      return checkObjectForToJsonProperty(object[key]);
+    }
+  });
+  return hasProperty;
+}
+
+function normalizeObject(object) {
+  if (checkObjectForToJsonProperty(object)) {
+    throw new Error('toJSON cannot be used as a property name');
+  }
+  const normalizedObject = JSON.parse(JSON.stringify(object));
+  ensureNoUnsafeProperties(normalizedObject);
+  return normalizedObject;
+}
+
+function normalizeString(string) {
+  return string ? String(normalizeObject(string)) : undefined;
```

Comment:
Sorry I might have misunderstood the comment, the reason I am checking if the string variable is defined, is because there are cases (for example when I am normalizing the index `const normalizedIndex = normalizeString(index);`) that the index is undefined. In that case the `String(normalizeObject(string))` will fail

Reply:
Sorry for being vague! What I meant is that `normalizeString` sounds like a function that's supposed to only work with strings and would reject arguments of any other type (via returning `undefined`). If that's the case then it seems there is no reason to use `normalizeObject(...)` and even `String(...)` if argument isn't a string __unless__ your goal is to also coerce numbers and booleans to strings (`true -> "true"`, `2 -> "2"`).

By the way if you call `normalizeObject(...)` with a string argument, then `checkObjectForToJsonProperty` (based on the current implementation) will iterate over all characters of the string (`Object.keys("string") -> ["s", "t", ....]`).

Reply:
Oh yes good point. As this has been added by Joe I need his feedback first. @flash1293 I see that we are using the normalizeString for the `index` and `alias`. Both index and alias should be strings so it makes sense here to check with `typeof string === 'string' ? string : undefined`

But we are also using it as a fallback for normalizeDate

```
function normalizeDate(date) {
  let normalizedDate = '';
  if (typeof date === 'number') {
    normalizedDate = date;
  } else if (date instanceof Date) {
    normalizedDate = date;
  } else {
    normalizedDate = normalizeString(date);
  }
  return normalizedDate;
}\```

Maybe here instead of `normalizeString` we can use `normalizeObject` and proceed with @azasypkin 's suggestion?